astroid/inference.py

```diff
@@ -1063,20 +1077,26 @@ def _cached_generator(func, instance, args, kwargs, _cache={}):  # noqa: B006
 # are mutated with a new instance of the property. This is why we cache the result
 # of the function's inference.
 @_cached_generator
-def infer_functiondef(self, context=None):
+def infer_functiondef(
+    self: nodes.FunctionDef, context: Optional[InferenceContext] = None
```

Comment:
`nodes.FunctionDef` is not quite enough since we also yield the same type. We need to use a bound TypeVar here.

Reply:
https://github.com/PyCQA/astroid/blob/393b27660f63c5c93d6e3bdb23da3534065564b2/astroid/inference.py#L1084

This is set to be a method on `FunctionDef`. It doesn't get used anywhere else. Can `self` be anything else than `FunctionDef`?

Reply:
What about `AsyncFunctionDef`? That inherits from `FunctionDef`.

Reply:
If it would only be used to annotate `self`, that wouldn't be an issue. However, here we also yield `self`. The result would be that `AsyncFunctionDef._infer` would yield just a `FunctionDef` which isn't correct.

Reply:
Pff, you must get tired of me making these easily avoidable mistakes. For real though, I truly appreciated the patience you're always showing with me. I'm learning so much from all the feedback you're giving me on my PRs. Very, very much appreciated! 😃 

Changed!